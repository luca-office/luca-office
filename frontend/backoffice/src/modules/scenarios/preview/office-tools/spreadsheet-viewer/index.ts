export * from "./use-spreadsheet-viewer-state"
export * from "./use-spreadsheet-viewer-state-actions"
export * from "./use-spreadsheet-viewer-survey-events"
