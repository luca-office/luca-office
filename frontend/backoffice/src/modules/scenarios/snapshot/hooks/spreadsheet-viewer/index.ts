export * from "./use-spreadsheet-viewer-snapshot-state"
export * from "./use-spreadsheet-viewer-snapshot-state-actions"
export * from "./use-spreadsheet-viewer-snapshot-survey-events"
