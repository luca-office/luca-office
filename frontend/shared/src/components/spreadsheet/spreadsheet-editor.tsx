import {css} from "@emotion/react"
import * as React from "react"
import {useRef} from "react"
import {HeadingLevel, IconName} from "../../enums"
import {SpreadsheetCellType} from "../../graphql/generated/globalTypes"
import {CellIndex, CellRange, SerializedCell, Spreadsheet} from "../../models"
import {
  backgroundColorLight,
  CustomStyle,
  Flex,
  spacing,
  spacingMedium,
  spacingSmall,
  spacingTinier,
  TextSize
} from "../../styles"
import {useLucaTranslation} from "../../translations"
import {Button, Heading, Icon, indexToCellName, Text, Tooltip} from ".."
import {DhxSpreadsheet, SpreadsheetApi, SpreadsheetStyleChange} from "./dhx-spreadsheet/dhx-spreadsheet"
import {
  defaultSpreadsheetColumnsCount,
  defaultSpreadsheetRowsCount,
  getFormulaExamples,
  getSheetDimensions
} from "./dhx-spreadsheet/dhx-utils"

export interface SpreadsheetEditorProps extends CustomStyle {
  readonly spreadsheet: Spreadsheet
  readonly readonly: boolean
  readonly rowsCount?: number
  readonly columnsCount?: number
  readonly isSaveLoading?: boolean
  readonly onCellValueChange?: (rowIndex: number, columnIndex: number, value: string) => void
  readonly onCellTypeChange?: (
    rowIndex: number,
    columnIndex: number,
    cellType: SpreadsheetCellType,
    value: string
  ) => void
  readonly onCellStyleChange?: (changes: Array<SpreadsheetStyleChange>) => void
  readonly onSelectCell?: (index: CellIndex) => void
  readonly onSelectCellRange?: (range: CellRange) => void
  readonly onSaveButtonClick?: (serializedCells: SerializedCell[]) => void
  readonly renderCustomFooter?: (spreadsheet: Spreadsheet, onSelectCell: (index: CellIndex) => void) => JSX.Element
}

export const SpreadsheetEditor: React.FC<SpreadsheetEditorProps> = props => {
  const {t} = useLucaTranslation()
  const {
    columnsCount: defaultColumnsCount = defaultSpreadsheetColumnsCount,
    customStyles,
    readonly,
    rowsCount: defaultRowsCount = defaultSpreadsheetRowsCount,
    spreadsheet,
    isSaveLoading,
    onCellValueChange,
    onCellTypeChange,
    onSelectCell,
    onSelectCellRange,
    onSaveButtonClick,
    renderCustomFooter,
    onCellStyleChange
  } = props
  const spreadsheetApiRef = useRef<SpreadsheetApi | null>(null)
  const [selectedCellName, updateSelectedCellName] = React.useState("A1")
  const {rowsCount, columnsCount} = getSheetDimensions(spreadsheet.cells, defaultColumnsCount, defaultRowsCount)

  const onSelectCellHandler = (index: CellIndex) => {
    updateSelectedCellName(indexToCellName(index))
    onSelectCell?.(index)
  }

  const saveButtonFooter =
    onSaveButtonClick !== undefined ? (
      <div css={styles.footer}>
        <Button
          isLoading={isSaveLoading}
          onClick={() => {
            const serializedCells = spreadsheetApiRef.current?.serialize()
            onSaveButtonClick(serializedCells ?? [])
          }}>
          {t("spreadsheet_editor__save_button")}
        </Button>
      </div>
    ) : null

  return (
    <div css={[styles.wrapper, customStyles]}>
      <DhxSpreadsheet
        spreadsheet={spreadsheet}
        rowsCount={rowsCount}
        columnsCount={columnsCount}
        readonly={readonly}
        customStyles={styles.dhxSpreadsheet}
        onCellValueChange={onCellValueChange}
        onCellTypeChange={onCellTypeChange}
        onSelectCell={onSelectCellHandler}
        onSelectCellRange={onSelectCellRange}
        onCellStyleChange={onCellStyleChange}
        onInit={spreadsheetApi => (spreadsheetApiRef.current = spreadsheetApi)}
        t={t}
      />
      {!readonly && (
        <>
          <div css={styles.selectedCellView}>
            <Text size={TextSize.Small} key={selectedCellName}>
              {selectedCellName}
            </Text>
          </div>
          <div css={styles.tooltipView}>
            <Tooltip
              title=""
              customTooltipStyles={styles.tooltip}
              placement="left"
              renderCustomContent={() => (
                <>
                  <Heading level={HeadingLevel.h3} customStyles={styles.tooltipHeader}>
                    {t("spreadsheet_editor__error_alert_text")}
                  </Heading>
                  <div css={styles.tooltipContentExamples}>
                    {getFormulaExamples().map((example, index) => (
                      <div key={index}>{example}</div>
                    ))}
                  </div>
                </>
              )}>
              <Icon customStyles={styles.tooltipViewIcon} name={IconName.Information} />
            </Tooltip>
          </div>
        </>
      )}
      {renderCustomFooter?.(spreadsheet, onSelectCellHandler)}
      {onSaveButtonClick !== undefined ? saveButtonFooter : null}
    </div>
  )
}

const Sizes = {
  selectionIndicatorWidth: 120
}

const styles = {
  wrapper: css({
    position: "relative",
    display: "flex",
    flexDirection: "column",
    width: "100%",
    height: "80vh"
  }),
  dhxSpreadsheet: css({
    width: "100%",
    height: "100%",
    flex: 1,

    ".editLine_wrapper": {
      marginLeft: Sizes.selectionIndicatorWidth - 1, // -1 to clear out the border
      width: `calc(100% - ${Sizes.selectionIndicatorWidth}px)`
    },

    ".dhx_header-fixed-cols": {
      zIndex: 21 // default was 999 -> problem with our modal, this is the smallest number possible
    }
  }),
  selectedCellView: css(Flex.row, {
    position: "absolute",
    justifyContent: "center",
    top: 45,
    left: 0,
    width: 120,
    height: 32,
    background: backgroundColorLight,
    boxShadow: "inset 0px 0px 4px 0px rgba(0, 0, 0, 0.16)"
  }),
  tooltipView: css(Flex.row, {
    position: "absolute",
    justifyContent: "center",
    top: 4,
    right: 4,
    width: 32,
    height: 38
  }),
  tooltipViewIcon: css({
    width: 18,
    height: 18
  }),
  tooltip: css({
    maxWidth: 450
  }),
  tooltipHeader: css({
    color: "white",
    marginBottom: spacingTinier,
    whiteSpace: "initial"
  }),
  tooltipContentExamples: css(Flex.column, {
    flexWrap: "wrap",
    marginTop: spacingSmall,
    maxHeight: 150
  }),
  footer: css({
    display: "flex",
    justifyContent: "flex-end",
    padding: spacing(spacingSmall, spacingMedium),
    backgroundColor: backgroundColorLight
  })
}
