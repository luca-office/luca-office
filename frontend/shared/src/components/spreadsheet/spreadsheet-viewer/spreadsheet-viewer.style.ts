import {css} from "@emotion/react"
import {Flex, spacingLarge, spacingMedium, spacingSmall} from "../../../styles"

export const spreadsheetViewerStyles = {
  viewer: css(Flex.column),
  editor: css({
    flex: 1,
    display: "flex",
    height: "initial"
  }),
  binary: css({
    flex: 1
  }),
  select: css({
    width: "100%"
  }),
  selectionFooterWrapper: css({
    display: "grid",
    gridTemplateColumns: "3fr 1fr 1fr 1fr",
    gap: spacingSmall,
    marginTop: spacingSmall,
    alignItems: "end",
    marginBottom: spacingSmall
  }),
  confirmButton: css({
    marginTop: spacingLarge,
    width: "100%"
  }),
  cellInput: css({
    marginTop: spacingMedium
  })
}
