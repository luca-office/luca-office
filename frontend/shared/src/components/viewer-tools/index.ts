export * from "./tools-header/tools-header"
export * from "./viewer-tools-sub-header/viewer-tools-sub-header"
