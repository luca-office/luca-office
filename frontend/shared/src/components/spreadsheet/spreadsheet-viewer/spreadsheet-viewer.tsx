import * as React from "react"
import {ViewerToolsType} from "../../../enums"
import {SpreadsheetCellType} from "../../../graphql/generated/globalTypes"
import {Binary, CellIndex, CellRange, SerializedCell, Spreadsheet, ViewerSpreadsheet} from "../../../models"
import {CustomStyle} from "../../../styles"
import {Option} from "../../../utils"
import {ContentLoadingIndicator} from "../../content-loading-indicator/content-loading-indicator"
import {ToolsHeader, ViewerToolsSubHeader} from "../../viewer-tools"
import {SpreadsheetStyleChange} from "../dhx-spreadsheet/dhx-spreadsheet"
import {DhxCellStyle} from "../dhx-spreadsheet/dhx-types"
import {SpreadsheetEditor} from "../spreadsheet-editor"
import {spreadsheetViewerStyles} from "./spreadsheet-viewer.style"

export interface SpreadsheetViewerProps extends CustomStyle {
  readonly activeSpreadsheetId?: UUID
  readonly isLoading?: boolean
  readonly onCloseSpreadsheet?: (id: UUID) => void
  readonly onCloseViewer: () => void
  readonly onMinimize?: () => void
  readonly readonly: boolean
  readonly setActiveSpreadsheetId?: (id: UUID) => void
  readonly spreadsheets: ViewerSpreadsheet[]
  readonly onCellValueChange?: (rowIndex: number, columnIndex: number, value: string) => void
  readonly onCellTypeChange?: (
    rowIndex: number,
    columnIndex: number,
    cellType: SpreadsheetCellType,
    value: string
  ) => void
  readonly onSelectCell?: (index: CellIndex) => void
  readonly onSelectCellRange?: (range: CellRange) => void
  readonly onCellStyleChange?: (changes: Array<SpreadsheetStyleChange>) => void
  readonly onSaveButtonClick?: (serializedCells: SerializedCell[]) => void
  readonly isSaveLoading?: boolean
  readonly renderCustomFooter?: (spreadsheet: Spreadsheet) => JSX.Element
}

export const SpreadsheetViewer: React.FC<SpreadsheetViewerProps> = props => {
  const {
    activeSpreadsheetId,
    customStyles,
    isLoading = false,
    onCloseSpreadsheet,
    onCloseViewer,
    onMinimize,
    readonly,
    setActiveSpreadsheetId,
    spreadsheets,
    renderCustomFooter,
    isSaveLoading,
    onCellValueChange,
    onCellTypeChange,
    onSelectCell,
    onSelectCellRange,
    onSaveButtonClick,
    onCellStyleChange
  } = props
  const selectedSheet = Option.of<ViewerSpreadsheet>(
    activeSpreadsheetId
      ? spreadsheets.find(sheet => sheet.id === activeSpreadsheetId)
      : spreadsheets[spreadsheets.length - 1]
  )

  const areControlsDisabled = spreadsheets.length <= 1 || !setActiveSpreadsheetId
  const selectedIndex = selectedSheet.map(binary => spreadsheets.indexOf(binary)).getOrElse(spreadsheets.length - 1)

  const onLeftClick = () => {
    if (setActiveSpreadsheetId && spreadsheets.length) {
      if (selectedIndex === 0) {
        setActiveSpreadsheetId(spreadsheets[spreadsheets.length - 1].id)
      } else {
        setActiveSpreadsheetId(spreadsheets[selectedIndex - 1].id)
      }
    }
  }

  const onRightClick = () => {
    if (setActiveSpreadsheetId && spreadsheets.length) {
      if (selectedIndex === spreadsheets.length - 1) {
        setActiveSpreadsheetId(spreadsheets[0].id)
      } else {
        setActiveSpreadsheetId(spreadsheets[selectedIndex + 1].id)
      }
    }
  }

  const mapSheetToBinary = (sheet: ViewerSpreadsheet) => ({id: sheet.id, title: sheet.title, path: ""} as Binary)

  const handleClose = () => {
    if (onCloseSpreadsheet !== undefined) {
      spreadsheets.forEach(spreadsheet => onCloseSpreadsheet(spreadsheet.id))
    }
    onCloseViewer()
  }

  return selectedSheet
    .map(spreadsheet => (
      <div css={[spreadsheetViewerStyles.viewer, customStyles]}>
        <ToolsHeader toolType={ViewerToolsType.TableEditor} onClose={handleClose} onMinimizeTool={onMinimize} />
        <ViewerToolsSubHeader
          elements={Option.of(spreadsheets.map(mapSheetToBinary))}
          selectedElement={selectedSheet.map(mapSheetToBinary)}
          closeElement={onCloseSpreadsheet}
          selectElement={setActiveSpreadsheetId}
          navigateToPrevious={areControlsDisabled ? undefined : onLeftClick}
          navigateToNext={areControlsDisabled ? undefined : onRightClick}
        />
        {isLoading ? (
          <ContentLoadingIndicator />
        ) : (
          <SpreadsheetEditor
            customStyles={spreadsheetViewerStyles.editor}
            spreadsheet={spreadsheet}
            readonly={readonly}
            onCellValueChange={onCellValueChange}
            onCellTypeChange={onCellTypeChange}
            onSelectCell={onSelectCell}
            onSelectCellRange={onSelectCellRange}
            onSaveButtonClick={onSaveButtonClick}
            isSaveLoading={isSaveLoading}
            renderCustomFooter={renderCustomFooter}
            onCellStyleChange={onCellStyleChange}
          />
        )}
      </div>
    ))
    .orNull()
}
