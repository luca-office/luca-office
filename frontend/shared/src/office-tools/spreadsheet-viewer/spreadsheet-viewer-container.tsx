/* eslint-disable max-lines */
import {keyBy} from "lodash-es"
import * as React from "react"
import {useEffect} from "react"
import {SpreadsheetViewer} from "../../components"
import {
  SpreadsheetStyleChange,
  SpreadsheetStyleChangeWithId
} from "../../components/spreadsheet/dhx-spreadsheet/dhx-spreadsheet"
import {SpreadsheetCellType} from "../../graphql/generated/globalTypes"
import {useFilesForSampleCompany, useFilesForScenario} from "../../graphql/hooks"
import {useSpreadsheetLazy} from "../../graphql/hooks/queries/use-spreadsheet-lazy"
import {
  CellIndex,
  CellRange,
  LocalSpreadsheet,
  SelectSpreadsheetCellEventPayload,
  SelectSpreadsheetCellRangeEventPayload,
  SelectSpreadsheetEventPayload,
  Spreadsheet,
  SpreadsheetCell,
  UpdateCellStyleSurveyEventPayload,
  UpdateCellTypeSurveyEventPayload,
  UpdateCellValueSurveyEventPayload,
  ViewerSpreadsheet
} from "../../models"
import {SpreadsheetState} from "../../redux/state/data"
import {CustomStyle} from "../../styles"
import {createUUID, find, getFileBySpreadsheetId, mergeRemoteFilesWithLocalSpreadsheets, Option} from "../../utils"

export interface SpreadsheetViewerState {
  readonly spreadsheets: Array<ViewerSpreadsheet>
  readonly selectedSpreadsheetId: Option<UUID>
  readonly localSpreadsheets: SpreadsheetState
  readonly isReadOnly: boolean
}

export interface SpreadsheetViewerStateActions {
  readonly initializeSpreadsheet: (spreadsheetId: UUID, spreadsheet: LocalSpreadsheet) => void
  readonly selectSpreadsheet: (spreadsheetId: UUID) => void
  readonly closeSpreadsheet: (spreadsheetId: UUID) => void
  readonly createValue: (spreadsheetId: UUID, cell: SpreadsheetCell) => void
  readonly changeValue: (spreadsheetId: UUID, cellId: UUID, value: string) => void
  readonly changeType: (spreadsheetId: UUID, cellId: UUID, value: SpreadsheetCellType) => void
  readonly changeStyleBatched: (spreadsheetId: UUID, changes: Array<SpreadsheetStyleChangeWithId>) => void
  readonly createCellBatched: (SpreadsheetId: UUID, cells: Array<SpreadsheetCell>) => void
}

export interface SpreadsheetViewerSurveyEvents {
  readonly sendSelectCellEvent: (payload: SelectSpreadsheetCellEventPayload) => void
  readonly sendRangeSelectionEvent: (payload: SelectSpreadsheetCellRangeEventPayload) => void
  readonly sendUpdateCellValueEvent: (eventProps: UpdateCellValueSurveyEventPayload) => void
  readonly sendUpdateCellTypeEvent: (eventProps: UpdateCellTypeSurveyEventPayload) => void
  readonly sendCloseSpreadsheetEvent: (spreadsheetId: UUID) => void
  readonly sendUpdateCellStyleEvent: (payload: UpdateCellStyleSurveyEventPayload) => void
  readonly sendSelectSpreadsheetEvent: (payload: SelectSpreadsheetEventPayload) => void
}

interface Props extends CustomStyle {
  readonly scenarioId: UUID
  readonly sampleCompanyIdOption: Option<UUID>
  readonly onClose: () => void
  readonly onMinimize?: () => void
  readonly useState: () => SpreadsheetViewerState
  readonly useStateActions: () => SpreadsheetViewerStateActions
  readonly useSurveyEvents: () => SpreadsheetViewerSurveyEvents
}

export const SpreadsheetViewerContainer: React.FC<Props> = ({
  customStyles,
  scenarioId,
  sampleCompanyIdOption,
  onClose,
  onMinimize,
  useState,
  useStateActions,
  useSurveyEvents
}) => {
  const {localSpreadsheets, selectedSpreadsheetId, spreadsheets, isReadOnly} = useState()

  const {
    closeSpreadsheet,
    initializeSpreadsheet,
    selectSpreadsheet,
    changeValue,
    createValue,
    changeType,
    changeStyleBatched,
    createCellBatched
  } = useStateActions()

  const {
    sendUpdateCellValueEvent,
    sendUpdateCellTypeEvent,
    sendSelectCellEvent,
    sendRangeSelectionEvent,
    sendCloseSpreadsheetEvent,
    sendUpdateCellStyleEvent,
    sendSelectSpreadsheetEvent
  } = useSurveyEvents()

  const onSpreadsheetLoaded = (spreadsheetOption: Option<Spreadsheet>) =>
    spreadsheetOption.forEach(spreadsheet =>
      initializeSpreadsheet(spreadsheet.id, {...spreadsheet, cells: keyBy(spreadsheet.cells, "id")})
    )

  const {spreadsheetLoading, getSpreadsheet} = useSpreadsheetLazy("cache-and-network", onSpreadsheetLoaded)
  const {files: filesForScenarioOption, filesLoading: filesForScenarioLoading} = useFilesForScenario(scenarioId)
  const {files: filesForSampleCompanyOption, filesLoading: filesForSampleCompanyLoading} = useFilesForSampleCompany(
    sampleCompanyIdOption.getOrElse(""),
    sampleCompanyIdOption.isEmpty()
  )

  useEffect(() => {
    selectedSpreadsheetId.forEach(id => localSpreadsheets[id] === undefined && getSpreadsheet(id))
  }, [selectedSpreadsheetId.orNull()])

  const files = React.useMemo(
    () =>
      mergeRemoteFilesWithLocalSpreadsheets(localSpreadsheets, [
        ...filesForScenarioOption.getOrElse([]),
        ...filesForSampleCompanyOption.getOrElse([])
      ]),
    [localSpreadsheets, filesForScenarioOption.orNull(), filesForSampleCompanyOption.orNull()]
  )

  const selectedSpreadsheet = find(spreadsheet => spreadsheet.id === selectedSpreadsheetId.orNull(), spreadsheets)

  const onCellTypeChange = (rowIndex: number, columnIndex: number, cellType: SpreadsheetCellType, value: string) =>
    selectedSpreadsheet.forEach(spreadsheet => {
      const spreadsheetId = spreadsheet.id
      const fileId = getFileBySpreadsheetId(files, spreadsheetId)
        .map(file => file.id)
        .orNull()
      const existingCell = spreadsheet.cells.find(
        cell => cell.rowIndex === rowIndex && cell.columnIndex === columnIndex
      )

      if (existingCell !== undefined && existingCell.cellType !== cellType) {
        changeType(spreadsheetId, existingCell.id, cellType)

        if (fileId !== null) {
          sendUpdateCellTypeEvent({
            fileId,
            spreadsheetId,
            rowIndex: existingCell.rowIndex,
            columnIndex: existingCell.columnIndex,
            cellId: existingCell.id,
            value,
            cellType
          })
        }
      }
    })

  const onValueChange = (rowIndex: number, columnIndex: number, value: string) =>
    selectedSpreadsheet.forEach(spreadsheet => {
      const spreadsheetId = spreadsheet.id
      const fileId = getFileBySpreadsheetId(files, spreadsheetId)
        .map(file => file.id)
        .orNull()
      const existingCell = spreadsheet.cells.find(
        cell => cell.rowIndex === rowIndex && cell.columnIndex === columnIndex
      )

      if (existingCell !== undefined) {
        changeValue(spreadsheetId, existingCell.id, value)

        if (fileId !== null) {
          sendUpdateCellValueEvent({
            fileId,
            spreadsheetId: existingCell.spreadsheetId,
            rowIndex: existingCell.rowIndex,
            columnIndex: existingCell.columnIndex,
            cellId: existingCell.id,
            cellType: existingCell.cellType,
            value
          })
        }
      } else {
        const cellId = createUUID()
        const cell: SpreadsheetCell = {
          __typename: "SpreadsheetCell",
          id: cellId,
          spreadsheetId,
          cellType: SpreadsheetCellType.General,
          rowIndex,
          columnIndex,
          value: value,
          style: null
        }

        createValue(spreadsheetId, cell)

        if (fileId !== null) {
          sendUpdateCellValueEvent({
            fileId,
            spreadsheetId: cell.spreadsheetId,
            rowIndex: cell.rowIndex,
            columnIndex: cell.columnIndex,
            cellId,
            cellType: cell.cellType,
            value
          })
        }
      }
    })

  const onSelectCell = ({rowIndex, columnIndex}: CellIndex) =>
    selectedSpreadsheet.forEach(spreadsheet => {
      find(cell => cell.rowIndex === rowIndex && cell.columnIndex === columnIndex, spreadsheet.cells).map(
        existingCell => {
          getFileBySpreadsheetId(files, spreadsheet.id)
            .map(file => file.id)
            .forEach(fileId => {
              sendSelectCellEvent({
                cellId: existingCell.id,
                columnIndex,
                rowIndex,
                fileId,
                spreadsheetId: spreadsheet.id
              })
            })
        }
      )
    })

  const onSelectCellRange = (range: CellRange) =>
    selectedSpreadsheet.forEach(spreadsheet => {
      getFileBySpreadsheetId(files, spreadsheet.id)
        .map(file => file.id)
        .forEach(fileId => {
          sendRangeSelectionEvent({
            startCellRowIndex: range.start.rowIndex,
            startCellColumnIndex: range.start.columnIndex,
            endCellRowIndex: range.end.rowIndex,
            endCellColumnIndex: range.end.columnIndex,
            fileId,
            spreadsheetId: spreadsheet.id
          })
        })
    })

  const onCellStyleChange = (changes: Array<SpreadsheetStyleChange>) => {
    selectedSpreadsheet.forEach(spreadsheet => {
      const fileId = getFileBySpreadsheetId(files, spreadsheet.id)
        .map(file => file.id)
        .orNull()

      if (!fileId) return

      const changeMap = changes.reduce((cellsMap, change) => {
        return {...cellsMap, [`${change.rowIndex}:${change.columnIndex}`]: change}
      }, {} as Record<string, SpreadsheetStyleChange>)

      const existingCellsToUpdate = spreadsheet.cells.filter(
        cell => changeMap[`${cell.rowIndex}:${cell.columnIndex}`] !== undefined
      )

      existingCellsToUpdate.forEach(cell => {
        sendUpdateCellStyleEvent({
          cellId: cell.id,
          columnIndex: cell.columnIndex,
          rowIndex: cell.rowIndex,
          style: changeMap[`${cell.rowIndex}:${cell.columnIndex}`].style,
          fileId,
          spreadsheetId: spreadsheet.id
        })
      })

      const cellMap = spreadsheet.cells.reduce((cellMap, cell) => {
        return {...cellMap, [`${cell.rowIndex}:${cell.columnIndex}`]: cell}
      }, {} as Record<string, SpreadsheetCell>)

      const cellUpdates = changes
        .filter(change => cellMap[`${change.rowIndex}:${change.columnIndex}`] !== undefined)
        .map(change => ({...change, id: cellMap[`${change.rowIndex}:${change.columnIndex}`].id}))

      changeStyleBatched(spreadsheet.id, cellUpdates)

      const cellCreations: Array<SpreadsheetCell> = changes
        .filter(change => cellMap[`${change.rowIndex}:${change.columnIndex}`] === undefined)
        .map(change => {
          const cellId = createUUID()

          // not very clean to have a side effect inside a map function, but ist soo convenient here!
          sendUpdateCellStyleEvent({
            cellId,
            columnIndex: change.columnIndex,
            rowIndex: change.rowIndex,
            style: change.style,
            fileId,
            spreadsheetId: spreadsheet.id
          })

          return {
            __typename: "SpreadsheetCell",
            id: cellId,
            spreadsheetId: spreadsheet.id,
            cellType: SpreadsheetCellType.General,
            rowIndex: change.rowIndex,
            columnIndex: change.columnIndex,
            value: "",
            style: JSON.stringify(change.style)
          }
        })

      createCellBatched(spreadsheet.id, cellCreations)
    })
  }

  const onCloseSpreadsheet = (spreadsheetId: UUID) => {
    closeSpreadsheet(spreadsheetId)
    sendCloseSpreadsheetEvent(spreadsheetId)
  }

  useEffect(() => {
    selectedSpreadsheetId.forEach(id => {
      const spreadsheet = spreadsheets.find(sheet => sheet.id === id)
      const fileOption = getFileBySpreadsheetId(files, id)

      fileOption.forEach(file => {
        if (spreadsheet !== undefined) {
          sendSelectSpreadsheetEvent({
            spreadsheetId: spreadsheet.id,
            spreadsheetTitle: spreadsheet.title,
            scenarioId,
            fileId: file.id,
            directoryId: file.directoryId
          })
        }
      })
    })
  }, [selectedSpreadsheetId.orNull()])

  return (
    <SpreadsheetViewer
      readonly={isReadOnly}
      isLoading={filesForSampleCompanyLoading || filesForScenarioLoading || spreadsheetLoading}
      customStyles={customStyles}
      onMinimize={onMinimize}
      onCloseViewer={onClose}
      onCloseSpreadsheet={onCloseSpreadsheet}
      activeSpreadsheetId={selectedSpreadsheetId.orUndefined()}
      setActiveSpreadsheetId={selectSpreadsheet}
      spreadsheets={spreadsheets}
      onCellValueChange={onValueChange}
      onCellTypeChange={onCellTypeChange}
      onSelectCell={onSelectCell}
      onSelectCellRange={onSelectCellRange}
      onCellStyleChange={onCellStyleChange}
    />
  )
}
