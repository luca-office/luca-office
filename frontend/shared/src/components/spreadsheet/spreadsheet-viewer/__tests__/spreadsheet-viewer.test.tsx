import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {ToolsHeader, ViewerToolsSubHeader} from "../../../viewer-tools"
import {spreadsheetsMock} from "../../__mocks__/spreadsheets.mock"
import {SpreadsheetEditor} from "../../spreadsheet-editor"
import {SpreadsheetViewer, SpreadsheetViewerProps} from "../spreadsheet-viewer"

const defaultProps: SpreadsheetViewerProps = {
  readonly: false,
  spreadsheets: spreadsheetsMock,
  onCloseViewer: jest.fn()
}

jest.mock("../../dhx-spreadsheet/dhx-spreadsheet", () => ({
  DhxSpreadsheet: () => <div>mockedSheet</div>
}))

const getComponent = (props?: Partial<SpreadsheetViewerProps>) => <SpreadsheetViewer {...{...defaultProps, ...props}} />

describe("spreadsheet-viewer", () => {
  it("renders correctly", () => {
    const component = getComponent()
    const tree = create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", () => {
    const component = getComponent()
    const tree = shallow(component)

    expect(tree.find(ToolsHeader)).toHaveLength(1)
    expect(tree.find(ViewerToolsSubHeader)).toHaveLength(1)
    expect(tree.find(SpreadsheetEditor)).toHaveLength(1)
  })
})
