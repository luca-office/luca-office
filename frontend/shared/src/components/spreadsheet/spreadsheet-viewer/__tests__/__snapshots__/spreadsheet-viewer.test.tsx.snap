// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spreadsheet-viewer renders correctly 1`] = `
<div
  className="css-kie12x-SpreadsheetViewer"
>
  <div
    className="css-1sdua8l-ToolsHeader"
  >
    <div
      className="css-w2t5un-ToolsHeader"
    >
      <div
        className="css-1ss9q1q-Icon"
        dangerouslySetInnerHTML={
          {
            "__html": {},
          }
        }
      />
      <div
        className="css-1b4s2bm-Text"
      >
        viewer_tools__calc_type_label
      </div>
    </div>
    <div
      className="css-w2t5un-ToolsHeader"
    >
      <button
        className="css-1dzgoe2-Button"
        disabled={false}
        onClick={[Function]}
        title="viewer_tools__close_tool_label"
      >
        <div
          className="css-95j2hm-Button"
        >
          <div
            className="css-13zef26-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
        </div>
      </button>
    </div>
  </div>
  <div
    className="css-2npy2c-ViewerToolsSubHeader"
  >
    <div
      className="css-14mtheo-ViewerToolsSubHeader"
    >
      <div
        className="css-1l2wmx6-ViewerToolsSubHeader"
      >
        <div
          className="css-0"
          id="tooltip_42"
        >
          <div
            className="subheader-tab css-1kbyu9y-SubheaderTab"
            data-scroll="subheader-tab_1233-abdc-5fer-grfs"
            data-testid="subheader-tab"
          >
            <div
              className="css-1sbf56p-Text"
            >
              default sheet
            </div>
          </div>
        </div>
        <div
          className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje css-10zyuev-Tooltip"
          role="tooltip"
          style={
            {
              "backgroundColor": "#2e3032",
              "zIndex": 1,
            }
          }
        >
          default sheet
          <div
            className="react-tooltip-arrow styles-module_arrow__K0L3T"
            style={{}}
          />
        </div>
        <div
          className="css-0"
          id="tooltip_42"
        >
          <div
            className="subheader-tab css-5kgfax-SubheaderTab"
            data-scroll="subheader-tab_1233-abdc-5fer-grfd"
            data-testid="subheader-tab"
          >
            <div
              className="css-1sbf56p-Text"
            >
              diverse cols and rows
            </div>
          </div>
        </div>
        <div
          className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje css-10zyuev-Tooltip"
          role="tooltip"
          style={
            {
              "backgroundColor": "#2e3032",
              "zIndex": 1,
            }
          }
        >
          diverse cols and rows
          <div
            className="react-tooltip-arrow styles-module_arrow__K0L3T"
            style={{}}
          />
        </div>
      </div>
    </div>
    <div
      className="css-rjlf5o-ViewerToolsSubHeader"
    >
      <button
        className="css-7pkjwh-Button"
        disabled={true}
        onClick={[Function]}
      >
        <div
          className="css-95j2hm-Button"
        >
          <div
            className="css-mj8vo1-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
        </div>
      </button>
      <button
        className="css-7pkjwh-Button"
        disabled={true}
        onClick={[Function]}
      >
        <div
          className="css-95j2hm-Button"
        >
          <div
            className="css-mj8vo1-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
        </div>
      </button>
    </div>
  </div>
  <div
    className="css-1gbwsr-SpreadsheetEditor"
  >
    <div>
      mockedSheet
    </div>
    <div
      className="css-vq7gv7-SpreadsheetEditor"
    >
      <div
        className="css-1uyodls-Text"
      >
        A1
      </div>
    </div>
    <div
      className="css-1q1k4o7-SpreadsheetEditor"
    >
      <div
        className="css-0"
        id="tooltip_42"
      >
        <div
          className="css-t0h4w2-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
      </div>
      <div
        className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje css-tewr92-Tooltip"
        role="tooltip"
        style={
          {
            "backgroundColor": "#2e3032",
            "zIndex": 1,
          }
        }
      >
        <div
          className="css-jb7kz7-Heading"
        >
          spreadsheet_editor__error_alert_text
        </div>
        <div
          className="css-1ovab5i-Tooltip"
        >
          <div>
            SUMME - SUM
          </div>
          <div>
            PRODUKT - PRODUCT
          </div>
          <div>
            MITTELWERT - AVERAGE/AVG
          </div>
          <div>
            ANZAHL - COUNT
          </div>
          <div>
            ANZAHL2 - COUNTA
          </div>
          <div>
            ZÃ„HLENWENN - COUNTIF
          </div>
          <div>
            SUMMEWENN - SUMIF
          </div>
          <div>
            WENN - IF
          </div>
          <div>
            SVERWEIS - VLOOKUP
          </div>
          <div>
            RUNDEN - ROUND
          </div>
          <div>
            MODUS.EINF - MODE.SNGL
          </div>
          <div>
            WURZEL - SQRT
          </div>
          <div>
            POTENZ - POWER
          </div>
        </div>
        <div
          className="react-tooltip-arrow styles-module_arrow__K0L3T"
          style={{}}
        />
      </div>
    </div>
  </div>
</div>
`;
